sympy/core/function.py

```diff
@@ -138,6 +138,19 @@ def __init__(cls, *args, **kwargs):
         cls._nargs = nargs
 
     @property
+    def __signature__(self):
+        """
+        Allow Python 3's inspect.signature to give a useful signature for
+        Function subclasses.
+        """
+        # Python 3 only, but this method is also only ever called in Python
+        # 3.
```

Comment:
Since IPython includes a backport of inspect.signature on Python 2, this statement isn't true. It will behave better if you wrap this import in 

```python
        try:
            from inspect import signature
        except ImportError:
            return None
```